### PR TITLE
WHALE_ALERT & COINGECKO new feeds

### PR DESCRIPTION
**Pull request by yohplala**
### Description of code

- [x] - Tested
- [x] - Changelog updated
- [x] - Contributors file updated (optional)

Hi @bmoscon 
This PR contains 4 main features:
- **1/** new feed `COINGECKO` coming with a new channel `profile`
- **2/** new feed `WHALE_ALERT` coming with a new channel `transactions` (on-chain transactions)
  Because `WHALE_ALERT` connects to a REST API requiring an API key, I implemented 2 ways to provide this API key:
  - **3/** either by writing it in `feed_keys.yaml` configuration file located in `cryptofeed` folder. Use of this file most notably allows querying the `info()` class method (even to query supported coins, an API key is required)
  - or by providing it as a parameter when initializing this feed. this 2nd use case is (IMHO) to be favored, and is most notably used when querying Whale Alert through cryptostore.
  Additional parameters (depending of the user pricing plan) are also required.
  All of them can (optionally) be provided when creating Whale Alert feed.
- **4/** strategy as proposed in #323 to optionally take into account the rate limit of the REST feed within the retry strategy impelmented in `FeedHandler` class.

I am removing the other PR that only contains CoinGecko feature.
Bests,

### CODE REVIEW

**Comment by bmoscon** [comment]
@yohplala can you resolve the conflicts?

File: README.md [hunk 1]
```diff
@@ -103,18 +108,24 @@ fh.add_nbbo([Coinbase, Bitfinex, HitBTC], ['BTC-USD'], nbbo_ticker)
 fh.run()
 ```
 
-## Supported Channels
```
Comment (bmoscon) [inline README.md:106]: revert this change

File: cryptofeed/feed.py [hunk 2]
```diff
@@ -4,21 +4,27 @@
...
 from cryptofeed.callback import Callback
-from cryptofeed.defines import (ASK, BID, BOOK_DELTA, FUNDING, L2_BOOK, L3_BOOK,
-                                LIQUIDATIONS, OPEN_INTEREST, TICKER, TRADES, VOLUME)
+from cryptofeed.defines import (ASK, BID, BOOK_DELTA, FUNDING, L2_BOOK, L3_BOOK, LIQUIDATIONS,
+                                OPEN_INTEREST, PROFILE, TICKER, TRADES, TRANSACTIONS, VOLUME)
 from cryptofeed.exceptions import BidAskOverlapping, UnsupportedDataFeed
 from cryptofeed.standards import feed_to_exchange, get_exchange_info, load_exchange_pair_mapping, pair_std_to_exchange
 from cryptofeed.util.book import book_delta, depth
 
+path = os.path.dirname(os.path.abspath(__file__))
```
Comment (bmoscon) [inline cryptofeed/feed.py:19]: constant type defs at the file level like this should be in all caps

File: cryptofeed/standards.py [hunk 3]
```diff
@@ -42,8 +44,8 @@ def load_exchange_pair_mapping(exchange: str):
             _std_trading_pairs[std] = {exchange: exch}
 
 
-def get_exchange_info(exchange: str):
-    mapping = gen_pairs(exchange)
+def get_exchange_info(exchange: str, key_id: str):
```
Comment (bmoscon) [inline cryptofeed/standards.py:47]: why did this change?

File: cryptofeed/standards.py [hunk 4]
```diff
@@ -30,10 +32,10 @@
 _exchange_to_std = {}
 
 
-def load_exchange_pair_mapping(exchange: str):
+def load_exchange_pair_mapping(exchange: str, key_id: str):
```
Comment (bmoscon) [inline cryptofeed/standards.py:35]: why did this change?

**Review by bmoscon — CHANGES_REQUESTED** [review]

**Comment by bmoscon** [comment]
a few other things - I think the name of the yaml file with the keys should be different, maybe just call it `config.yaml` so it can be used for other things. Also the config should be at the feedhandler level, not a feed level. feedhandler can load the config and various parts of the system can use it as applicable.

**Comment by yohplala** [comment]
> @yohplala can you resolve the conflicts?

Hi @bmoscon, conflicts solved.

File: README.md (hunk 1 above)
Comment (yohplala) [reply README.md:106]: Ok, change reverted.

File: cryptofeed/feed.py (hunk 2 above)
Comment (yohplala) [reply cryptofeed/feed.py:19]: @bmoscon, this is part of the 'config file' implementation topic: I comment on this topic below.

File: cryptofeed/standards.py (hunk 4 above)
Comment (yohplala) [reply cryptofeed/standards.py:34]: Idem as above
WHALE_ALERT feed required an API key even to query recognized coins.
So I need to forward the API key to all functions that call `gen_pairs()`.
This is the case for `load_exchange_pair_mapping`.

File: cryptofeed/standards.py (hunk 3 above)
Comment (yohplala) [reply cryptofeed/standards.py:46]: WHALE_ALERT feed required an API key even to query recognized coins.
So I need to forward the API key to all functions that call `gen_pairs()`.
This is the case for `get_exchange_info`.

> a few other things - I think the name of the yaml file with the keys should be different, maybe just call it `config.yaml` so it can be used for other things. Also the config should be at the feedhandler level, not a feed level. feedhandler can load the config and various parts of the system can use it as applicable.


Bryant,
The sole purpose of `feed_keys.yaml` is to provide to the feed requiring it (at the moment only WHALE_ALERT) a `key` class parameter without requiring creation of an object.
The sole need is to be able to call the class method `info` without more parameter.
If we move this config file at `FeedHanlder` level, then this need is not satisfied. `key` cannot be made a class parameter of `Feed` class.

In this case, the other approach to have `info` class method working (that I just realize now) is simply to add a `key_id` parameter to that function:
```python
@classmethod
def info(cls, key_id: str = None) -> dict:
```

I can do that, but then I don't need any longer a config file.
So, if you agree, if you feel that a config file for `Feed` class is not appropriate (and I have no trouble with that, no worries), I will simply remove it, and do as proposed above: add a `key_id` parameter to `info` class method.

Would this be ok?
(adding a config file at `FeedHandler` level could then be managed in a separate PR, as a separate feature)

Just let me know what is you preference, and I will adjust accordingly.
Bests,

**Comment by yohplala** [comment]
> a few other things - I think the name of the yaml file with the keys should be different, maybe just call it `config.yaml` so it can be used for other things. Also the config should be at the feedhandler level, not a feed level. feedhandler can load the config and various parts of the system can use it as applicable.


Bryant,
The sole purpose of `feed_keys.yaml` is to provide to the feed requiring it (at the moment only WHALE_ALERT) a `key` class parameter without requiring creation of an object.
The sole need is to be able to call the class method `info` without more parameter.
If we move this config file at `FeedHanlder` level, then this need is not satisfied. `key` cannot be made a class parameter of `Feed` class.

In this case, the other approach to have `info` class method working (that I just realize now) is simply to add a `key_id` parameter to that function:
```python
@classmethod
def info(cls, key_id: str = None) -> dict:
```

I can do that, but then I don't need any longer a config file.
So, if you agree, if you feel that a config file for `Feed` class is not appropriate (and I have no trouble with that, no worries), I will simply remove it, and do as proposed above: add a `key_id` parameter to `info` class method.

Would this be ok?
(adding a config file at `FeedHandler` level could then be managed in a separate PR, as a separate feature)

Just let me know what is you preference, and I will adjust accordingly.
Bests,

**Comment by bmoscon** [comment]
info() could still load out the key from a config file based on the exchange the info method is invoked on. If you want to wait for my code changes that use exchange keys thats fine too, but its at the feedhandler level, and the config file lives in the project root.

**Comment by yohplala** [comment]
> info() could still load out the key from a config file based on the exchange the info method is invoked on. If you want to wait for my code changes that use exchange keys thats fine too, but its at the feedhandler level, and the config file lives in the project root.

@bmoscon 
Bryant, please, can you clarify how you would then want to forward the key to `info()`? As a parameter? (i.e. key would not be an attribute of the `Feed` instance, right?)
(to make sure we are aligned on that)

Bryant, I will align with the way you see it, no worries. I just need to know how you see this variable given to the `info()` method.

Also, I have to 'suspend' this PR, I am sorry. I realize this evening that my Whale Alert feed is not recording since a day, while I can see alerts on their TG channel. Something is wrong, and I have to investigate.
I am really sorry, you will feel it becomes an habit, ah ah.
I will analyze this as soon as possible.

**Comment by yohplala** [comment]
Hi @bmoscon 
To keep you informed on this topic:
- I think I have understood what was being wrong in my code / why I did not retrieve all transactions. I did not put enough margin delay when querying back in time for 'new' transactions.
- I have further open a discussion with support to make sure the strategy I implemented does not leave chance to miss transactions. I.e. I am waiting confirmation that older transactions can not pop up in their db after newer transaction (older and newer being base on the timestamp of the transactions when confirmed in blockchain, and not time of being inserted in the db)

Well, I think, I got it right now, but waiting confirmation from support to make sure it is.

During the week, I will also remove the config file stuff to leave you room to implement it the way you see it fit.
And it should be good for next week-end I think.
Bests,

**Comment by yohplala** [comment]
Hi @bmoscon 
So current status as of today:
- I have rolled out my corrections in Whale Alert, it seems ok. Let's wait till end of this week to see if it still is.
- Meanwhile, I have also removed the Feed local config file as I proposed.

Bryant, if you have some time, please, can you review the PR while it is currently running?
I confirm the branch is currently updated, and if I don't have any more problem by end of this week, I will not touch anything more, except upon your request of course.

Thanks for your support, have a good evening,
Bests,

**Comment by yohplala** [comment]
Hi @bmoscon 
I have pushed a new version, which I hope will be the last. I am starting this evening a new monitoring phase :)
In the end, there were 3 bugs: 2 in `whale_alert.py` and 1 in `feedhandler.py` (`delay` and `retry` not resetted)
In `whale_alert.py`, after different exchanges with support, I have noticeably simplified the query method.
all should be fixed now.
I will let you know next week-end how it is going.

**Comment by warproxxx** [comment]
Hey just wanted to say thank you for working on this wonderful addition :)

File: README.md [hunk 5]
```diff
@@ -106,16 +111,22 @@ fh.run()
...
 * VOLUME
 * FUNDING
 * BOOK_DELTA - Subscribed to with L2 or L3 books, receive book deltas rather than the entire book on updates. Full updates will be periodically sent on the L2 or L3 channel. If BOOK_DELTA is enabled, only L2 or L3 book can be enabled, not both. To receive both create two `feedhandler` objects. Not all exchanges are supported, as some exchanges send complete books on every update.
-* OPEN_INTEREST - Open interest data
+* OPEN_INTEREST - Open interest data.
+
+Aggregated data from provider is available in channel:
+
+* TRANSACTIONS - On-chain transactions.
+* PROFILE - current aggregated price, market cap, volume (in USD, BTC or ETH currency), total and circulating supply,
```
Comment (bmoscon) [inline README.md:128]: I'm trying to think of a better name than profile. Profile seems like the name that would be used for private account information. The information here is more summary or generic statistics. What about INFORMATIONAL? or something like that, any thoughts?

File: cryptofeed/provider/coingecko.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    async def _profile(self, session, pair):
+        """
+        Data from /coins/{id}.
+        """
+
+        quote_c, base_c = pair.split('_')
+
+        async with session.get(f"{self.address}coins/{quote_c}?localization=false&tickers=false\
+&market_data=true&community_data=true&developer_data=false&sparkline=false") as response:
```
Comment (bmoscon) [inline cryptofeed/provider/coingecko.py:77]: can you make this a single line?

File: cryptofeed/provider/whale_alert.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+from json import JSONDecodeError
+
+from cryptofeed.defines import WHALE_ALERT, TRANSACTIONS
+from cryptofeed.feed import RestFeed
+from cryptofeed.standards import pair_exchange_to_std
+from cryptofeed.exceptions import RestResponseError
+
+
+LOG = logging.getLogger('feedhandler')
+to_from_data = ('address', 'owner_type', 'owner')
```
Comment (bmoscon) [inline cryptofeed/provider/whale_alert.py:21]: can't you define this in the method its used in? why is this global?

**Review by bmoscon — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (yohplala) [reply README.md:128]: Hi @bmoscon 
I am fine with INFORMATIONAL if this is ok, english-ly speaking :)  I am french, so don't ask me which word could be more suited for this :)
Instead of PROFILE, it could also be COIN_PROFILE
Coingecko calls this MARKET, but I am not fond of that.
Within this data, there are sentiment data as well.
INFORMATIONAL makes sense yes, why not.

Comment (bmoscon) [reply README.md:128]: I'm fine with MARKET, MARKET_INFO, or COIN_PROFILE - take your pick :)

Comment (yohplala) [reply README.md:128]: @bmoscon, so let this be MARKET_INFO ! :)

File: cryptofeed/provider/whale_alert.py (hunk 7 above)
Comment (yohplala) [reply cryptofeed/provider/whale_alert.py:21]: Yes, I can define it directly in the method using it.
Well, I am kind of always caring about CPU performance (always, like really always :)), and this data is absolutely not dynamic. So once the source is read, python has it, once and for all the execution of the code (this is my understanding).
If I implement it where it is used (in `_transactions()`), it would then be declared for every iteration, every coin, over and over again...
I will do as you propose Bryant, but I am curious to know what is the drawback at keeping it global within `whale_alert.py`?

Comment (bmoscon) [reply cryptofeed/provider/whale_alert.py:21]: if you want it in the global scope then please make it capitalized, TO_FROM_DATA. Also, pretty sure it would be faster with a set than a generator for what you are doing. The iterable is only 3 elements, a generator is overkill

Comment (yohplala) [reply cryptofeed/provider/whale_alert.py:21]: Bryant!
It is not a generator, it is a tuple. A tuple is event faster to declare than a list.

Comment (yohplala) [reply cryptofeed/provider/whale_alert.py:21]: > if you want it in the global scope then please make it capitalized, TO_FROM_DATA.

Ok, doing right away!

File: README.md (hunk 5 above)
Comment (yohplala) [reply README.md]: ok, done!

File: cryptofeed/provider/coingecko.py (hunk 6 above)
Comment (yohplala) [reply cryptofeed/provider/coingecko.py]: ok, done!

**Comment by yohplala** [comment]
@bmoscon,
I have updated the source accordingly.
I will update you on how it is running on my side tomorrow or the day after and if everything is ok, I think PR is ready.

**Review by bmoscon — APPROVED** [review]

**Comment by bmoscon** [comment]
@yohplala looks good to merge, just let me know when you've re-tested and I'll merge it

**Comment by yohplala** [comment]
Hi @bmoscon ,
It is my very pleasure to confirm that my last check is ok.
For me, good to go! :)
